### PR TITLE
Syntax highlighting for JSM

### LINKED ISSUE #2155: .JSM syntax highlighting
.jsm files (regular javascript) are still used in many addons for Firefox. It would be nice if Git Extensions could highlight syntax of these files. Github is highlighting the files.

I don't see any option for enable this.

### PR DESCRIPTION
**Pull request by drewnoakes**
Fixes #2155

Changes proposed in this pull request:
- Add syntax highlighting for `.jsm` files
- ~~~Make the loading of `AutoCompleteRegexes.txt` use async IO (this occurs during the first load of `FormCommit`)~~~

What did I do to test the code and ensure quality:
- Manual testing

Has been tested on:
- GIT 2.18
- Windows 10

### CODE REVIEW

File: GitUI/AutoCompletion/CommitAutoCompleteProvider.cs [hunk 1]
```diff
@@ -14,7 +14,10 @@ namespace GitUI.AutoCompletion
 {
     public class CommitAutoCompleteProvider : IAutoCompleteProvider
     {
-        private static readonly Lazy<Dictionary<string, Regex>> _regexes = new Lazy<Dictionary<string, Regex>>(ParseRegexes);
+#pragma warning disable VSTHRD012 // Provide JoinableTaskFactory where allowed
+        private static readonly AsyncLazy<Dictionary<string, Regex>> _regexByExtension = new AsyncLazy<Dictionary<string, Regex>>(ParseRegexesAsync);
+#pragma warning restore VSTHRD012 // Provide JoinableTaskFactory where allowed
```
Comment (drewnoakes) [inline GitUI/AutoCompletion/CommitAutoCompleteProvider.cs]: @sharwell could you verify that this diagnostic can be skipped in this scenario? The async operation only awaits disk IO operations so as I see it there's no opportunity for deadlock.

Getting a JTF in this static initialiser is awkward. We'd likely need a new `Initialise` instance method, or convert the ctor to a static factory.

File: GitUI/AutoCompletion/CommitAutoCompleteProvider.cs [hunk 2]
```diff
@@ -34,13 +37,15 @@ public CommitAutoCompleteProvider(GitModule module)
             {
                 cancellationToken.ThrowIfCancellationRequested();
 
-                var regex = GetRegexForExtension(Path.GetExtension(file.Name));
+                var extension = Path.GetExtension(file.Name);
+
+                var regexByExtension = await _regexByExtension.GetValueAsync(cancellationToken);
```
Comment (RussKie) [inline GitUI/AutoCompletion/CommitAutoCompleteProvider.cs]: does this need to be async?

Comment (drewnoakes) [reply GitUI/AutoCompletion/CommitAutoCompleteProvider.cs]: Previous code parsed file synchronously during the first call to `Lazy<Dictionary<...>>.Value`. That happend during show of `FormCommit`, which would block during IO, reading the file.

This change uses `AsyncLazy` which allows the factory function to be async, which allows async IO.

Comment (RussKie) [reply GitUI/AutoCompletion/CommitAutoCompleteProvider.cs]: we can probably initiate this process when we start the app, and thus won't need such complex structure, can probably replace it with `ConcurrentDictionary`...
thoughts?

Comment (drewnoakes) [reply GitUI/AutoCompletion/CommitAutoCompleteProvider.cs]: Could do. It's not a big saving here anyway as the file is small and fast to read. On the order of a few milliseconds in my testing with an SSD. I mostly wanted to experiment a little more with vs-threading.

Comment (drewnoakes) [reply GitUI/AutoCompletion/CommitAutoCompleteProvider.cs]: I'm happy to drop the second commit from this PR.

Comment (RussKie) [reply GitUI/AutoCompletion/CommitAutoCompleteProvider.cs]: KISS, if we can help it

**Comment by drewnoakes** [comment]
Peeled the last commit off.

**Comment by vbjay** [comment]
LOL!!!

See now that's just going to confuse us stupid people.  :D

Keep
It
Simple
Stupid

**Comment by RussKie** [comment]
Nice one :)))

**Pull request merged**